### PR TITLE
Bugfix/2241 rng model ctor

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submisison Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary
Stan model's 3-arg constructor takes unsigned int (RNG seed) instead of RNG object itself; change discard strategy for RNG inits accordingly.

#### Intended Effect
Move burden of choosing/instantiating the RNG object used by the model constructor from the interfaces to core Stan.

#### How to Verify
Rung unit test src/test/unit/lang/generator_test.cpp

#### Side Effects
The function stan/services/util/create_rng.hpp use the chain id directly as the stride (discard) instead of using (chain_id - 1).  Interfaces should continue to enforce chain id > 0; the Stan model constructor calls the function create_rng using chain id 0.

#### Documentation
None

#### Reviewer Suggestions
Bob Carpenter or Sean Talts or Daniel Lee

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Mitzi Morris

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/test/unit/model/finite_diff_grad_test.cpp [hunk 1]
```diff
@@ -92,7 +92,8 @@ TEST(ModelUtil, streams) {
   stan::io::dump data_var_context(data_stream);
   data_stream.close();
 
-  stan_model model(data_var_context, 0);
+  std::stringstream output;
+  stan_model model(data_var_context, &output);
```
Comment (seantalts) [inline src/test/unit/model/finite_diff_grad_test.cpp]: I think you can do this and keep the existing semantics here:
```
-  std::stringstream output;
-  stan_model model(data_var_context, &output);
+  stan_model model(data_var_context, (std::stringstream*)0);
```

File: src/stan/lang/generator/generate_constructor.hpp [hunk 2]
```diff
@@ -26,27 +26,28 @@ namespace stan {
...
       // body of constructor now in function
-      o << INDENT << "template <class RNG>" << EOL;
       o << INDENT << "void ctor_body(stan::io::var_context& context__," << EOL;
-      o << INDENT << "               RNG& base_rng__," << EOL;
+      o << INDENT << "               unsigned int random_seed__," << EOL;
       o << INDENT << "               std::ostream* pstream__) {" << EOL;
+      o << INDENT2 << "boost::ecuyer1988 base_rng__ =" << EOL;
+      o << INDENT2 << "  stan::services::util::create_rng(random_seed__, 0);"
+        << EOL;
+      o << INDENT2 << "(void) base_rng__;  // suppress unused var warning"
```
Comment (bob-carpenter) [inline src/stan/lang/generator/generate_constructor.hpp]: The user can call `_rng` functions in transformed data.

File: src/stan/services/util/create_rng.hpp [hunk 3]
```diff
@@ -22,7 +22,7 @@ namespace stan {
         // Advance generator to avoid process conflicts
         static boost::uintmax_t DISCARD_STRIDE
           = static_cast<boost::uintmax_t>(1) << 50;
-        rng.discard(DISCARD_STRIDE * (chain - 1));
+        rng.discard(DISCARD_STRIDE * (chain));
```
Comment (bob-carpenter) [inline src/stan/services/util/create_rng.hpp]: The -1 was there before to make sure chain 1 (lowest legal value in the interfaces) didn't advance the PRNG.  Now we're going to use the PRNG to generate transformed data and then advance chain 1 by 2^50 draws.

File: src/stan/lang/generator/generate_constructor.hpp (hunk 2 above)
Comment (bob-carpenter) [inline src/stan/lang/generator/generate_constructor.hpp]: Transformed data can have `_rng` functions.  I thought I'd already replied that, but I don't see it now.

**Comment by bob-carpenter** [comment]
I fixed the merge conflict, but this now failing here on my local tests:

```
test/unit/services/sample/hmc_nuts_dense_inv_metric --gtest_output="xml:test/unit/services/sample/hmc_nuts_dense_inv_metric.xml"
Running main() from gtest_main.cc
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from ServicesSampleHmcNutsDenseEMassMatrix
[ RUN      ] ServicesSampleHmcNutsDenseEMassMatrix.ident_no_adapt
[       OK ] ServicesSampleHmcNutsDenseEMassMatrix.ident_no_adapt (6 ms)
[ RUN      ] ServicesSampleHmcNutsDenseEMassMatrix.ident_adapt_250
src/test/unit/services/check_adaptation.hpp:63: Failure
The difference between param_vals[ij] and test::unit::stod(strs[j]) is 0.45182600000000006, which exceeds err_margin, where
param_vals[ij] evaluates to 1.2162200000000001,
test::unit::stod(strs[j]) evaluates to 0.76439400000000002, and
err_margin evaluates to 0.20000000000000001.
[  FAILED  ] ServicesSampleHmcNutsDenseEMassMatrix.ident_adapt_250 (103 ms)
[ RUN      ] ServicesSampleHmcNutsDenseEMassMatrix.use_metric_no_adapt
[       OK ] ServicesSampleHmcNutsDenseEMassMatrix.use_metric_no_adapt (1 ms)
[ RUN      ] ServicesSampleHmcNutsDenseEMassMatrix.use_metric_skip_adapt
[       OK ] ServicesSampleHmcNutsDenseEMassMatrix.use_metric_skip_adapt (1 ms)
[----------] 4 tests from ServicesSampleHmcNutsDenseEMassMatrix (111 ms total)
```

@mitzimorris and @betanalpha --- can you work this out?

**Comment by betanalpha** [comment]
Looks like the test assumed a particular input RNG seed which changes in this PR.

**Pull request merged**